### PR TITLE
Allow Processors.count == 0

### PR DESCRIPTION
**Pull request by jan-elastic**

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/unit/Processors.java [hunk 1]
```diff
@@ -137,7 +137,7 @@ public int roundDown() {
     }
 
     private static boolean validNumberOfProcessors(double processors) {
-        return Double.isFinite(processors) && processors > 0.0;
+        return Double.isFinite(processors) && processors >= 0.0;
```
Comment (jan-elastic) [inline server/src/main/java/org/elasticsearch/common/unit/Processors.java:140]: Judging by the rest of this file zero should be allowed.

E.g. 
- the constructor starts with `if (count == 0.0)` (line 42)
- the exception message reads `processors must be a non-negative number` (line 59)

Comment (jan-elastic) [reply server/src/main/java/org/elasticsearch/common/unit/Processors.java:140]: This is causing the `/_ml/info` endpoint to fail when no ML nodes are started.

See also (private Slack channel): https://elastic.slack.com/archives/G2FUG2B61/p1701184000628989

**Review by thecoop — APPROVED** [review]
The constructor can accept 0 processors, so the check method should too

**Comment by jan-elastic** [comment]
> The constructor can accept 0 processors, so the check method should too

Some more observations:
- the constructor is private; nothing outside this class can call it
- `Processor.ZERO` calls it with argument 0
- a few other tests are failing, suggesting this is intended behavior

I'll wait for others to chime in before I proceed with this and fix the broken tests.

**Comment by jan-elastic** [comment]
Maybe another method like `Processors.ofOrZero(double)` is a better (less invasive) fix.

Otherwise I have to construct a `Processors` in the ML code as:
```
total == 0 ? Processors.ZERO : Processors.of((double) total)
```
which looks a bit stupid in my opinion.

**Comment by fcofdez** [comment]
The issue here is that we're using `Processors` to represent the actual number of configured processors for the ES process (which cannot be 0) and to represent information about the required capacity for autoscaling (hence the `Processors#Zero` instance). In hindsight, I think we should have used a different class to represent the number of processors in `AutoscalingResources`. I would prefer to keep `Processors` as it is and move the check elsewhere.

**Comment by thecoop** [comment]
In which case does it make sense to remove the check for 0.0 in the constructor?

**Comment by jan-elastic** [comment]
I was already afraid this change would have too much consequences downstream. I'll close this PR.

**Comment by jan-elastic** [comment]
The constructor is called with 0 in line 30:
```
public static final Processors ZERO = new Processors(0.0);
```

**Pull request closed without merging**